Revise checkbounds again

timholy opened the pull request:
`checkbounds` was revised in #17137 and #17340, but the latter left a few dangling uncalled `_chkbnds` in `multidimensional.jl`. Here's another proposed revision that will hopefully be performant ~~(haven't had time to test yet)~~ and also be a cleaner & more general API. As a bonus, this version is the first (to my knowledge) to ever support arrays of `CartesianIndex`es, which will be important if we ever decide to make `find*` return such arrays. Finally, lots of tests added.

CC @JeffBezanson, @mbauman, @andreasnoack (ref https://github.com/JuliaParallel/DistributedArrays.jl/pull/74).

timholy committed:
Force inlining of `linearindices` and `indices1`.

Ref #17340

timholy committed:
Revise checkbounds API and extend to support Array{CartesianIndex}

timholy commented:
Might as well let @nanosoldier `runbenchmarks(ALL, vs = ":master")` do most of the work.

tkelman commented:
Does this help resolve the method errors for OneTo seen in several packages? Wasn't that intended to be mostly an internal detail?

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/17355#issuecomment-231560633) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/36f904066207070e15e804b4b82d6a6bc6d2e9b9/817bdec_vs_2065920/report.md). cc @jrevels

timholy commented:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/17355#issuecomment-231585257) has completed - no performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/34b0ccad2bfd20799707061ce530b4b00982313b/a21d09f_vs_f8d67f7/report.md). cc @jrevels

timholy commented:
Tests pass and benchmarks look great, so it's just a question of whether we want this.

tkelman commented on base/multidimensional.jl:
```diff
@@ -146,16 +146,41 @@ end  # IteratorsMD
+    checkindex(Bool, inds1, I[1]) & checkbounds_indices(indsrest, tail(I))
+end
+
+@inline split{N}(inds, V::Type{Val{N}}) = _split((), inds, V)
```
should this be inside the `IteratorsMD` module?

timholy replied:
If we don't think this is broadly useful, moving it would indeed be a good idea.

tkelman replied:
there is already a `split` function for strings, defining these methods in `Base` would be extending that same function but I think this is sufficiently different that it should be separate

mbauman commented on base/abstractarray.jl:
```diff
@@ -644,8 +662,9 @@ _getindex{T,N}(::LinearSlow, A::AbstractArray{T,N}, I::Vararg{Real, N}) = (@_pro
 function _getindex(::LinearSlow, A::AbstractArray, i::Real)
     # ind2sub requires all dimensions to be > 0; may as well just check bounds
     @_inline_meta
-    @boundscheck checkbounds(A, i)
```
Why not keep this as `checkbounds`? Won't this make custom behaviors harder to implement?

timholy replied:
That's how it was at first, but surprisingly LLVM doesn't generate very good code; there was a [10x performance regression]() despite the fact that everything is nominally inlined (these lines are the difference between the first and second benchmark reports).

I will try to dig into the underlying reasons further.

mbauman commented:
Could you add a little documentation on the intended purpose of all these functions?  As I understand it:

* `checkbounds(A, I...)`: allows `A` to implement its own bounds checking behaviors, but by default it calls:
* `checkbounds_indices(::Tuple{T1, …}, ::Tuple{T2, …})`, which allows `T2` to implement its own index expansion (e.g., for CartesianIndices where one element maps to more than one dimension). This guy is undocumented and not exported.  By default it iteratively calls:
* `checkindex(::Type{Bool}, ::T1, ::T2)`, which allows `T2` to define its own method for determining if its within the bounds of `T1`.  Is there a reason to preserve the `::Bool` first argument?

timholy commented:
Yes, you have the idea.

> This guy is undocumented and not exported

As far as not documented, I did provide [this comment](https://github.com/JuliaLang/julia/blob/f13bd541cb7061c9b8119e69914c9bb3ca07bc2c/base/abstractarray.jl#L207-L215). Is this insufficient? Would converting that into a docstring address your concern?

It definitely isn't exported, and I'm not yet convinced it should be. For almost all arrays it seems like an internal detail---as long as the array provides useful `indices`, I can't see why extending `checkbounds_indices` should be necessary in any case except for novel index types that consume ≠1 entry in the `indices`-tuple. However, I'm happy to change this if this seems wrong.

> Is there a reason to preserve the ::Bool first argument?

I think I did it for consistency with `checkbounds`, and it leaves room for a `checkindex(::T1, ::T2)` that throws when its conditions are not met, but I've never found a need to introduce this variant. I'm happy to drop the `Bool` if that seems preferable.

timholy commented:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

timholy commented:
OK, I think I've figured out what was happening: it was a lack of call-site specialization for the splatted `checkbounds`. By adding a 1-index specialization (eff2593b16ef851cfe61fb2841df3906f4e4cf4d), locally I recovered the performance. I'm not entirely happy with this solution, because it's one more definition to specialize for a custom array type, but I don't see a way around this (I even annotated `checkbounds` with `@pure`, but that didn't help; in contrast, annotating the caller, `getindex`, with `@pure` does fix the problem).

However, detailed investigation revealed something relatively extraordinary that I had not yet noticed, and while it's a little long (apologies) I think it's kind of fun:
```jl
include(Pkg.dir("BaseBenchmarks","src","array","sumindex.jl"))
a = rand(1000,1000)
b = ArrayLS(a)
s = sub(a, 1:999, 1:1000)  # LinearSlow
perf_sumlinear(a)
perf_sumlinear(b)
perf_sumlinear(s)
@time 1
@time perf_sumlinear(a)
@time perf_sumlinear(b)
@time perf_sumlinear(s)
```

julia-0.4:
```
  0.000002 seconds (148 allocations: 10.151 KB)
  0.001049 seconds (5 allocations: 176 bytes)
  0.016766 seconds (5 allocations: 176 bytes)
  0.035747 seconds (2.00 M allocations: 30.479 MB, 12.65% gc time)
```

master:
```
  0.000003 seconds (137 allocations: 8.438 KB)
  0.001039 seconds (5 allocations: 176 bytes)
  0.001045 seconds (5 allocations: 176 bytes)
  0.009655 seconds (5 allocations: 176 bytes)
```
If you look carefully, there's a >10x performance improvement in linear indexing for `ArrayLS`. That initially seems rather exciting. However, my guess is that what's happening is that there's enough inlining now that LLVM notices that it can elide the `div`, so we're unlikely to get that kind of performance boost in any situation in which we actually need linear indexing (as in the SubArray case). We can test this theory like this:
```jl
immutable ArrayLS1{T,N} <: MyArray{T,N}  # LinearSlow
    data::Array{T,N}
end

@inline Base.size{T}(A::ArrayLS1{T,2}) = (sz = size(A.data); (sz[1]-1,sz[2]-1)) # this line forces LLVM to compute the div

@inline Base.getindex(A::ArrayLS1, i::Int, i2::Int) = getindex(A.data, i, i2)
@inline Base.unsafe_getindex(A::ArrayLS1, i::Int, j::Int) = Base.unsafe_getindex(A.data, i, j)

c = ArrayLS1(a)
perf_sumlinear(c)
@time perf_sumlinear(c)
```
which yields julia-0.4:
```
  0.016899 seconds (5 allocations: 176 bytes)
```
and master:
```
  0.009288 seconds (5 allocations: 176 bytes)
```
So all that fighting I did against the 10x regression seems a little bit pointless, since it now looks like a (new) artifact of the benchmark. If we take that viewpoint seriously, should I drop eff2593b16ef851cfe61fb2841df3906f4e4cf4d, since I'd rather not have it? Or is it worth having for those corner cases where something is declared `LinearSlow` but in reality it's `LinearFast`?

Presumably it would make sense to revise the benchmarks to make them like ArrayLS1?

Of course, we're still doing considerably better than julia-0.4, including a 4x improvement for `SubArray`. So there's definitely some genuine good news here, but less dramatic than it initially seemed.

mbauman commented:
Really, what I'm after is a higher level strategy.  The call graphs are really deep… and the only reason is to simplify dispatch and allow for easier extensions by custom arrays/indices.  I actually missed a step above.

I don't think `checkbounds_indices` needs to be exported or officially supported, either.  I was just trying to describe how things work… and thinking about how to document the strategy here.

I don't feel strongly about `checkindex(::Bool, …)`.  I'm fine with the status quo… it's just really hard to change these guys since they're designed to be extended, so I want to limit future changes.

timholy commented:
> Really, what I'm after is a higher level strategy.

This is one of those cases where I confess I'm struggling with "author blindness" to what are doubtlessly holes in the documentation. To see if I understand, is your description in https://github.com/JuliaLang/julia/pull/17355#issuecomment-231622202 basically a summary of what you'd like in terms of documentation? (I'm happy to copy/paste/edit that into the source code.) Doesn't that comment I linked to above (conveniently [here again](https://github.com/JuliaLang/julia/blob/f13bd541cb7061c9b8119e69914c9bb3ca07bc2c/base/abstractarray.jl#L207-L215)) basically say the same thing? Would it be clearer if I described it like
```
    custom array type->take its indices tuple->call `checkindex` on matching pairs of indices
```
?

Or are you looking for "strategic vision"? This can be a little hard to articulate (and I should probably try harder), but here are a couple:
- `checkbounds`->`checkbounds_indices`: while we should (and do) provide an array-specific method (`checkbounds`), in general bounds-checking is about comparing the allowed indices against the requested indices. For that reason, it seems cleaner to phrase the main bounds-checking logic purely in terms of comparisons between a pair of indices (hence `checkbounds_indices`).
- by taking the `indices` tuple and getting rid of the specific array type at the first stage, we avoid needing to recompile the whole bounds-checking callgraph for each array type. (I suspect this is the main reason the subarray tests have gotten considerably faster of late.)
- pairing elements of tuples, rather than carrying around the array, seems to make it a tad easier to provide flexibility: it seems just a little easier to "advance the counter" selectively on one tuple or the other. For example, now we can do this:

```jl
julia> A = rand(3,4)
3×4 Array{Float64,2}:
 0.546929   0.092763  0.158707  0.405476
 0.0441324  0.536253  0.596574  0.79432 
 0.976893   0.760052  0.56792   0.396879

julia> I = [CartesianIndex((1,2)), CartesianIndex((3,4))]
2-element Array{CartesianIndex{2},1}:
 CartesianIndex{2}((1,2))
 CartesianIndex{2}((3,4))

julia> A[I]
2-element Array{Float64,1}:
 0.092763
 0.396879
```
whereas on julia-0.4 this yields
```jl
julia> A[I]
ERROR: ArgumentError: unable to check bounds for indices of type CartesianIndex{2}
 in getindex at abstractarray.jl:488
```

> The call graphs are really deep… and the only reason is to simplify dispatch and allow for easier extensions by custom arrays/indices.

Yes, this is a big bummer. If you see a way to avoid that design pattern, I'm all ears. I'll point out that I don't think it's any worse than `index_lengths`, `index_shape`, `reindex`, and others which you may be more familiar with. Heck, it's not exactly like `_internal_checkbounds` and related functions are particularly simple on julia-0.4: if I ditch eff2593b16ef851cfe61fb2841df3906f4e4cf4d (which I'm tempted to do for other reasons), delete all blank lines, comments, and docstrings, and don't count the 16 lines of code needed to support that `CartesianIndex` example above (which is a new feature), then bounds-checking is exactly 50 lines on both julia-0.4 and here. That doesn't mean there isn't a difference in complexity; I don't really see one, but I can't judge well since I'm recently the author of this one.

timholy committed:
Faster and simpler indices for SubArray

timholy committed:
Don't extend split to tuples (move it into IteratorsMD)

timholy commented:
Thanks for taking the time to look it over! I agree it's been a somewhat chaotic process.

I will copy/paste some version of your analysis at the top of the bounds-checking code in abstractarray.jl.

timholy commented:
Now documented out the wazoo. I moved some code around and renamed some variables for consistency from one function to the next, so there's more apparent than real churn. 

@nanosoldier `runbenchmarks(ALL, vs = ":master")`

timholy commented:
With nanosoldier feeling better, let's try again: @nanosoldier `runbenchmarks(ALL, vs = ":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/17355#issuecomment-232087923) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/9dcaa7b11fc1ab036c06ef34d488d55838e31d40/019ba50_vs_f881efb/report.md). cc @jrevels

timholy commented:
Seems like https://github.com/JuliaCI/BaseBenchmarks.jl/pull/16 wasn't used on the nanosoldier run; I've checked locally and can't replicate any of the regressions. I think this is ready to go.

tkelman commented on base/range.jl:
```diff
@@ -326,12 +326,17 @@ step(r::AbstractUnitRange) = 1
 step{T}(r::LinSpace{T}) = ifelse(r.len <= 0, convert(T,NaN), (r.stop-r.start)/r.divisor)
 
-function length(r::StepRange)
+unsafe_length(r::Range) = length(r)  # generic fallback
```
what is potentially unsafe about unsafe_length?

timholy replied:
See the `checked_sub` variants a little farther down in the file. The checked versions are necessary if you want to avoid being fooled when you do this:
```jl
julia> r = typemin(Int):typemax(Int)
-9223372036854775808:9223372036854775807

julia> length(r)
ERROR: OverflowError()
 in length(::UnitRange{Int64}) at ./range.jl:354
 in eval(::Module, ::Any) at ./boot.jl:234
 in macro expansion at ./REPL.jl:92 [inlined]
 in (::Base.REPL.##1#2{Base.REPL.REPLBackend})() at ./event.jl:46
```
The downside of checked arithmetic is it's god awful slow if you use it for something performance-critical like bounds checking.

Fortunately, since for arrays all sizes have to be >= 0, it's not a cost you need to pay. So it's worth having a fast path for carefully-chosen applications.

tkelman commented on base/abstractarray.jl:
```diff
@@ -144,77 +150,82 @@ linearindexing(::LinearFast, ::LinearFast) = LinearFast()
+There are two import exceptions to the 1-1 rule: linear indexing and
+CartesianIndex{N}, both of which may "consume" more than one element
+of `IA`.
 """
```
the docstring should be attached to the most general signature (or a zero-method line)

timholy commented:
Thanks for the review as always, @tkelman.

tkelman commented:
note that there are several sphinx warnings from the formatting in your rst

timholy committed:
Add a lot of documentation on the bounds checking hierarchy [ci skip]

This also reorganizes code to make the order follow the hierarchy; makes more sense for the "big picture" documentation to be near the top node.

timholy commented:
Hmm, I already encountered an application where it was really sweet to be able to call `checkbounds_indices` directly. (I'm creating a `ColorView` array type, which takes a 3-by-m-by-n `Array{T}` and acts like a m-by-n `RGB{T}` array---it was nice to be able to check bounds without needing the actual array itself.) The unexpected utility of this design materialized rather quickly!

For that reason, I decided I had better symmetrize everything: now all functions have both `foo(Bool, args...)` (returns a `Bool`) and `foo(args...)` (throws if out-of-bounds) variants.

tkelman commented on base/replutil.jl:
```diff
@@ -147,11 +147,13 @@ function showerror(io::IO, ex::BoundsError)
         if isa(ex.a, AbstractArray)
             print(io, summary(ex.a))
+        elseif isa(ex.a, Tuple)
+            print(io, "array with indices ", ex.a)
```
is a `BoundsError` with a `Tuple` field always going to be the indices?

timholy replied:
Amazing we didn't have a test for that. Added.

tkelman replied:
ah i see with your new version, you won't usually index with `i` a tuple

JeffBezanson commented on base/replutil.jl:
```diff
@@ -147,11 +147,13 @@ function showerror(io::IO, ex::BoundsError)
         print(io, ": attempt to access ")
         if isa(ex.a, AbstractArray)
             print(io, summary(ex.a))
+        elseif isa(ex.a, Tuple) && isdefined(ex, :i) && isa(ex.i, Tuple)
```
`ex.a` is also a Tuple when you index a tuple out of bounds.

tkelman replied:
`i` won't be though, see collapsed discussion right above

JeffBezanson replied:
Ah, I see. Technically unambiguous, but kind of obscure.

timholy replied:
I don't like the ambiguity either. Some options I see:
- create a new exception type (yuck)
- add another field to `BoundsError`
- create an internal `ArrayIndices` type that wraps the tuple-of-indices and sticks it in the `a` slot

Of these, I like the third the best.

timholy replied:
Option 4: don't support the throwing version of `checkbounds_indices`. Really I was most concerned about setting up an expectations that `checkbounds_indices(IA, I)` would throw, and so if that yields a `MethodError` but `checkbounds_indices(Bool, IA, I)` works, any user will expect that to return a `Bool` and then could handle the throwing him/herself. Maybe this is the best choice?

JeffBezanson replied:
Yes, option 4 sounds pretty good.

JeffBezanson commented:
I'm basically ok with this (especially since @mbauman is :) ). Overloading the meaning of the fields in BoundsError is kind of unfortunate --- won't that lead to unpredictable messages for bounds errors, where sometimes the array type is given and sometimes it isn't?

timholy committed:
Non-throwing boundscheck functions are written as f(Bool, args...)

timholy commented:
Test failure is #17409.

tkelman commented on base/abstractarray.jl:
```diff
@@ -144,6 +150,120 @@ linearindexing(::LinearFast, ::LinearFast) = LinearFast()
+    length(A) == length(I)
+end
+function checkbounds_logical(::Type{Bool}, A::AbstractVector, I::AbstractVector{Bool})
+    indices(A) == indices(I)
```
Are this and `checkbounds_logical(::Type{Bool}, A::AbstractVector, I::AbstractArray{Bool})` swapped? Should the `vector[array]` case be checking `indices`, and `vector[vector]` be checking length?

timholy replied:
As I'm thinking about it, no, but I could be missing something so I welcome discussion. The idea is that `A[i]` is retained if `I[i] == true`. Consequently, the indices of `A` and `I` have to match. The exception to the requirement for exact matching is when we're using linear indexing. But when both of them are vectors, linear indexing is trumped by Cartesian indexing.

tkelman replied:
Okay, makes sense. I'm not entirely sure what you mean by "linear indexing is trumped by Cartesian indexing" though.

timholy replied:
https://github.com/JuliaLang/julia/blob/3627ed27bb7084bb8b32c0e210694e6f632ecf1d/base/abstractarray.jl#L1242-L1245
